pennylane/math/sparse_matrix.py

```diff
@@ -0,0 +1,186 @@
+# Copyright 2018-2021 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Class for framework-agnostic sparse matrix representation."""
+from copy import copy
+import numpy as np
+import pennylane as qml
+
+
+class SparseMatrix:
+    """Framework-agnostic class to represent a sparse matrix.
+
+    The sparse matrix is stored as a dictionary of a tuple of indices as keys, and
+    the corresponding value as entries:
+
+    >>> s = SparseMatrix(tf.Variable([[3., 0., 0.],[2., 0., 0.]]))
+    >>> print(s.data)
+    {(0, 0): <tf.Tensor: shape=(), dtype=float32, numpy=3.0>, (1, 0): <tf.Tensor: shape=(), dtype=float32, numpy=2.0>}
+    >>> print(s.shape)
+    (2, 3)
+
+    An empty sparse matrix can be initialised by passing a shape tuple:
+
+    >>> s = SparseMatrix((12, 12))
+    >>> print(s.data)
+    None
+    >>> print(s.shape)
+    (12, 12)
+
+    Args:
+        arg (tuple or tensor_like): if tuple, this is interpreted as the shape of
+            the sparse matrix; else it is interpreted as a 2-d tensor which is a dense representation
+            of the matrix.
+
+    **Example:**
+
+    To mimic ``scipy.sparse`` objects, sparse matrices give acces to lists of row and column indices,
+    as well as the number of non-zero elements:
+
+    >>> s = SparseMatrix(tf.Variable([[3., 0.],[2., 0.]]))
+    >>> print(s.row)
+    [0, 1]
+    >>> print(s.col)
+    [0, 0]
+    >>> print(s.nnz)
+    2
+
+    They support elementary arithmetic such as addition, subtraction and scalar multiplication from both sides:
+
+    >>> s1 = SparseMatrix(torch.tensor([[3., 0.],[2., 0.]]))
+    >>> s2 = SparseMatrix(torch.tensor([[-3., 0.],[0., 0.]]))
+    >>> res = s1 + s2
+    >>> print(res.data)
+    {(1, 0): tensor(2.)}
+
+    >>> res2 = s1 - s2
+    >>> print(res2.data)
+    {(0, 0): tensor(6.), (1, 0): tensor(2.)}
+
+    >>> res3 = 2. * s1
+    >>> print(res3.data)
+    {(0, 0): tensor(6.), (1, 0): tensor(4.)}
+
+    We can also compute the Kronecker product (or tensor product) of two sparse matrices:
+    >>> res4 = s1.kron(s2)
+    >>> print(res4.data)
+    {(0, 0): -9.0, (2, 0): -6.0}
+    >>> print(res4.shape)
+    (4, 4)
+    """
+
+    def __init__(self, arg):
+
+        if isinstance(arg, tuple):
+            self.data = {}
+            self.shape = arg
+
+        else:
+            self.shape = qml.math.shape(arg)
+            self.data = {}
```

Comment:
If there is a danger that this gets mixed up with `ops.data` we should consider renaming. Note that `scipy.sparse.coo_matrix.data` refers to a list of elements!